File: adler8.pas
unit adler8;
//Adler-8
//Author: domasz
//Version: 0.1 (2022-11-17)
//Licence: MIT

interface

uses SysUtils, HasherBase;

type THasherAdler8 = class(THasherbase)
  private
    FHash, FHash2: Byte;
  public
    constructor Create; override;
    procedure Update(Msg: PByte; Length: Integer); override;
    function Final: String; override;
end;

implementation

constructor THasherAdler8.Create;
begin
  inherited Create;
  Check := '7A';
  FHash := 1;
  FHash2 := 0;
end;

procedure THasherAdler8.Update(Msg: PByte; Length: Integer);
const MOD_ADLER = 13;
var i: Integer;
begin
  for i:=0 to Length-1 do begin
    FHash := (FHash + Msg^) mod MOD_ADLER;
    FHash2 := (FHash2 + FHash) mod MOD_ADLER;
    Inc(Msg);
  end;   
end;

function THasherAdler8.Final: String;
begin
  FHash := (FHash2 shl 4) or FHash;
  Result := IntToHex(FHash, 2);
end;

initialization
  HasherList.RegisterHasher('Adler-8', THasherAdler8);

end.

File: PHP/Algos/adler8.php
<?php
//Adler-8
//Author: domasz
//Version: 0.1 (2022-11-19)
//Licence: MIT

require_once 'HasherBase.php';

class HasherAdler8 extends HasherBase
{
	protected $FHash, $FHash2 = 0;

	public function __construct() 
	{
		$this->Check = '7A';
		$this->FHash = 1;
		$this->FHash2 = 0;
	}

	public function Update($Msg, $Length)
	{
		$MOD_ADLER = 13;
		
		for ($i=0; $i<$Length; $i++)
		{
			$this->FHash = ($this->FHash + ord($Msg[$i]) ) % $MOD_ADLER;
			$this->FHash2 = ($this->FHash2 + $this->FHash) % $MOD_ADLER;
		}
	}

	public function Finish()
	{
		$this->FHash = ($this->FHash2 << 4) | $this->FHash;
		return sprintf('%02X', $this->FHash);
	}
}

$HasherList->RegisterHasher('Adler-8', 'HasherAdler8');